[BUG] test_casting_from_overflow_double_to_timestamp `DID NOT RAISE <class 'Exception'>`

**Describe the bug**
Tests failure on Databricks-11.3
```
[2024-02-15T16:06:00.629Z] [31mFAILED[0m ../../src/main/python/orc_cast_test.py::[1mtest_casting_from_overflow_double_to_timestamp[DATAGEN_SEED=1708012573][0m - Failed: DID NOT RAISE <class 'Exception'>
```

**Repro**

The test failure reproduces with an upstream OSS build as well:
```
DATAGEN_SEED=1708012573 \
TZ=UTC \
SPARK_HOME=~/dist/spark-3.5.0-bin-hadoop3 \
TEST_PARALLEL=0 \
./integration_tests/run_pyspark_from_build.sh \
-k test_casting_from_overflow_double_to_timestamp
```

<details>
<summary>Failed: DID NOT RAISE <class 'Exception'></summary>
<pre>
   @allow_non_gpu(*non_utc_allow_for_test_casting_from_overflow_long)
    def test_casting_from_overflow_double_to_timestamp(spark_tmp_path):
        orc_path = spark_tmp_path + '/orc_casting_from_overflow_double_to_timestamp'
        with_cpu_session(
            lambda spark: unary_op_df(spark, DoubleGen(min_exp=38)).write.orc(orc_path)
        )
>       assert_gpu_and_cpu_error(
            df_fun=lambda spark: spark.read.schema("a timestamp").orc(orc_path).collect(),
            conf={},
            error_message="ArithmeticException"
        )

../../src/main/python/orc_cast_test.py:137: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../src/main/python/asserts.py:661: in assert_gpu_and_cpu_error
    assert_spark_exception(lambda: with_gpu_session(df_fun, conf), error_message)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

func = <function assert_gpu_and_cpu_error.<locals>.<lambda> at 0x7face2933910>
error_message = 'ArithmeticException'

    def assert_spark_exception(func, error_message):
        """
        Assert that a specific Java exception is thrown
        :param func: a function to be verified
        :param error_message: a string such as the one produce by java.lang.Exception.toString
        :return: Assertion failure if no exception matching error_message has occurred.
        """
>       with pytest.raises(Exception) as excinfo:
E       Failed: DID NOT RAISE <class 'Exception'>

../../src/main/python/asserts.py:646: Failed
</pre>
</details>

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala
@@ -402,19 +402,15 @@ object GpuOrcScan {
         // In this step, ORC casting of CPU throw an exception rather than replace such values with
         // null. We followed the CPU code here.
         withResource(milliseconds) { _ =>
-          // Test whether if there is long-overflow
-          // If milliSeconds.max() * 1000 > LONG_MAX, then 'Math.multiplyExact' will
-          // throw an exception (as CPU code does).
+          // Test whether if there is long-overflow towards positive and negative infinity
           withResource(milliseconds.max()) { maxValue =>
-            if (maxValue.isValid) {
-              testLongMultiplicationOverflow(maxValue.getDouble.toLong,
-                DateTimeConstants.MICROS_PER_MILLIS)
-            }
-          }
-          withResource(milliseconds.min()) { minValue =>
-            if (minValue.isValid) {
-              testLongMultiplicationOverflow(minValue.getDouble.toLong,
-                DateTimeConstants.MICROS_PER_MILLIS)
+            withResource(milliseconds.min()) { minValue =>
+              Seq(maxValue, minValue).foreach { extremum =>
+                if (extremum.isValid) {
+                  testLongMultiplicationOverflow(extremum.getDouble.toLong,
+                    DateTimeConstants.MICROS_PER_MILLIS)
+                }
+              }
             }
           }
           withResource(Scalar.fromDouble(DateTimeConstants.MICROS_PER_MILLIS)) { thousand =>
@@ -492,7 +488,7 @@ object GpuOrcScan {
    * In Math.multiplyExact, if there is an integer-overflow, then it will throw an
    * ArithmeticException.
    */
-  def testLongMultiplicationOverflow(a: Long, b: Long) = {
+  private def testLongMultiplicationOverflow(a: Long, b: Long) = {
     Math.multiplyExact(a, b)
   }
 

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2019-2023, NVIDIA CORPORATION.
+ * Copyright (c) 2019-2024, NVIDIA CORPORATION.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -408,6 +408,12 @@ object GpuOrcScan {
           withResource(milliseconds.max()) { maxValue =>
             if (maxValue.isValid) {
               testLongMultiplicationOverflow(maxValue.getDouble.toLong,
+                DateTimeConstants.MICROS_PER_MILLIS)
+            }
+          }
+          withResource(milliseconds.min()) { minValue =>
+            if (minValue.isValid) {
+              testLongMultiplicationOverflow(minValue.getDouble.toLong,
                 DateTimeConstants.MICROS_PER_MILLIS)
             }
           }

